### PR TITLE
Add preliminary GitHub integration

### PR DESCRIPTION
**Pull request by Alfred-Mountfield**
## 🌟 What is the purpose of this PR?

This PR adds a lot of the foundation required to be able to ingest data from external sources. It mainly focuses on doing so by implementing a route using the [Github Airbyte Connector](https://github.com/airbytehq/airbyte/tree/master/airbyte-integrations/connectors/source-github).

This currently is able to be ran through the GraphQL APIs, allowing a (mostly) non-configurable ingest of Github data (although you can pick the repositories and timeframe to import over). It's implemented through a selection of routes on the hash-task-executor, which runs the connector's docker container in a subprocess. There is a route for the 4 commands a connector supports: `[Spec, Check, Discover, Read]`.

## ⚠️ Known issues
- Entity data can contain nested objects in arrays, which breaks the EntityType editor
- Ingest is very slow, and the whole stream is serialized into memory at each step
- The ConfiguredCatalog is not actually configurable, at the moment we select everything for ingest
- There is no support for incremental state (i.e. we do a full ingest every time and do not have de-duplication)
- The Github API token is stored in plain text in a temp folder and its removal isn't ensured
- No support for OAuth
- There are no inferred entityTypes (e.g. User entities that are nested inside other data)
- No Links are created on ingest
- The AirByte protocol types are currently manually generated from a file in their repo

## 🐾 Next steps

- Add changes to the front-end to allow the user to configure the streams the connector extracts
- Store the users API credentials somewhere
- Make types and routes available between integration implementations and the HASH api package
- Implement de-duplication logic (avoid re-ingesting existing data)
- Implement incremental state support
- Move away from reading the whole of stdout and HTTP requests to streams to allow ongoing processing of ingested data rather than needing to handle it all in memory in a single batch
- Add intermediary logging to give progress information
- Add more connectors
- Move away from the task-executor to something like Temporal

## 🔍 What does this change?

- Migrates the hash-task-executor to express
- Adds nodemon to the hash-task-executor for dev convenience
- Adds an auto-generated set of types created from [the Airbyte Protocol](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml)
- Adds an Airbyte executor interface which uses docker images and supports the 4 methods
  - Spec
  - Check
  - Discover
  - Read
- Implements a Github specific version of the Airbyte Executor
- Adds a somewhat _hacky_ `ExistingEntityChecker` which:
  - Queries all types associated with the user's accountId
  - Builds a hashmap of stream names to EntityTypeIds
  - Has a method to check if there's an EntityType in the DB already which has the same name as the stream name
  - Has a method to create a new EntityType given the stream name
- Adds 4 new graphQL routes for the methods outlined above, allowing the user to provide connector **configuration** but **not** its selection of streams
- Adds 4 resolvers, which in union with the task-executor implementation does:
  - Spec
    - Simply calls the task-executor route
    - Returns the spec of the connector
  -  Check
    - Takes a configuration that contains credentials to check connector authentication
    - Returns a connection status
  - Discover
    - Takes a configuration that contains credentials for authentication
    -  **Creates a `catalog.json` file in the local code directory**, configuring it for selecting all streams
    - Task-executor returns the catalog (not configuredCatalog)
    - API backend uses the `ExistingEntityChecker` to create new EntityTypes for all streams that don't already have EntityTypes
  - Read
    - Takes a configuration that contains credentials to check connector authentication
    - **Reads the `catalog.json` file in the local code directory**
    - Uses the `ExistingEntityChecker` to find the EntityTypes associated with the stream, and inserts entities

### 📜 Does this require a change to the docs?

- This is still something not intended to be used outside of very specific integrations development, and until the time it's used more widely in the team, docs are misleading and add additional overhead.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1202055186269230/1202140499897843/f) _(internal)_
- [AirByte Connectors](https://docs.airbyte.com/connector-development/)
- [Github API Spec](https://docs.github.com/en/rest)

## 🛡 What tests cover this?

- No automated ones

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Run external services and `seed-data`
1. Run `yarn dev`
1. Run `yarn workspace @hashintel/hash-task-executor start`
1. Open the front-end and login to generate the appropriate cookies
1. Open the GraphQL playground (http://localhost:5001/graphql)
1. Try each of the following mutations in order:
   ```graphql
   mutation {
     executeGithubSpecTask
   }
   ```
   ```graphql
   mutation {
     executeGithubCheckTask(config:{
       repository: "blockprotocol/blockprotocol"
       start_date: "2022-04-21T00:00:00Z"
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```
   ```graphql
   mutation {
     executeGithubDiscoverTask(config:{
       repository: "blockprotocol/blockprotocol"
       start_date: "2022-04-21T00:00:00Z"
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```
   ```graphql
   mutation {
     executeGithubReadTask(config:{
       repository: "blockprotocol/blockprotocol"
       start_date: "2022-04-21T00:00:00Z"
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```

## 📹 Demo
image
image
image
image
image
image
image

### CODE REVIEW

File: packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:105]: This whole object is pretty messy, not sure what we want to do with it. Can merge it with a `@todo` or something but I think generally our API isn't well laid out right now to find an EntityType like we need to here.

It's also based on some assumptions about accountId which I don't think is good.

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:105]: Absolutely, just opened the PR to get it viewed earlier in the process. Agreed on comments if we merge this as is

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:105]: Agreed on that too, this was mostly a hack to get my development forwards, and I opened the other comment here to check what we want to do with it, wasn't expecting us to merge it as is. I suspect we instead want to actually add a method on the EntityType class (just realised this should be `ExistingEntityTypeChecker` anyway), _perhaps_ even before merging this PR

**Comment by Alfred-Mountfield** [comment]
Thanks @thehabbos007 for the review, I'll come back to them next week when we resume normal processes :)

File: packages/hash/api/src/graphql/resolvers/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/index.ts:147]: Nevermind, have reverted on this decision due to casing concerns, especially if we end up automatically converting between cases (as in from camel to title for example) using a utility.

File: packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:121]: Right but I don't understand why the entityTypeId _on an EntityType_ is on the `entityId` field? 

Is it because `entityId` is an alias for `id` and I'm therefore actually calling `entityType.id` (which makes sense) and just being mislead?

Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:121]: I think we moved away from `id` as it lead to confusion, and we want to specify what id we're referring to from any type. 
This instance of `id` could just have been an oversight.

If I recall correctly, `entityTypeId` actually refers to 'the' `EntityType`  system type when resolved in the GQL API. That makes it so that the fields on the `EntityType` model class aren't called `entityType(Version)Id`.
This system type also referred to as the `EntityTypeType` in some places. I find this generally confusing.

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:121]: I agree with all of that, I'm just still uncertain as to what we're actually discussing as an action item here. The state of things right now is that on an `EntityType`(e.g. the objects returned by `getAccountEntityTypes`) you have:
* `entityType.entityTypeId := undefined`
* `entityType.entityId := [the entityType ID]`

This clearly isn't right, I don't even know why the `EntityType` type has an `entityId` or `entityVersionId` property as it's not an entity.

File: packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:105]: Do we want to add a `getByAccountIdAndName` method or something on the `EntityType` class?

File: packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:121]: This question was more of an observation when first writing.
As an action item, I would like for us to capture these issues in an asana task (there might be one already?), as they are overlapping, and since it does lead to some confusing debugging sessions while writing code around EntityTypes.

File: packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:114]: Good question, I'm not sure why the `null` is here. I think it must have been to do with satisfying some type constraints or something somewhere that is now redundant. Will remove

File: packages/hash/api/src/task-execution/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/task-execution/index.ts:30]: My Rust habits still letting me down here

Comment (thehabbos007) [reply packages/hash/api/src/task-execution/index.ts:30]: long live snake_case

File: packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:121]: Captured in (asana task)[https://app.asana.com/0/1200211978612931/1200809642643269/f] (_internal_)

File: packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:105]: Sorry, didn't see this response -
I think it might take a little more work, as the return value of this function is a little unconventional compared to the rest of the methods in the model class. 

While the consistency would be nice, I think moving it to `api/src/task-executor` would be fine for now, and then we can move it into the model class down the line. For me the thin resolvers are more important in this case

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts]: Makes sense, moved it

File: packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts]: Cheers, I've updated the todo to include a link to it

**Review by thehabbos007 — APPROVED** [review]
Looking nice! I am approving this with a couple comments. 
I would suggest not merging until Ben/Ciaran has had a chance to look at it too.

File: packages/hash/api/src/graphql/typeDefs/taskExecution.typedef.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/taskExecution.typedef.ts:8]: I'm trying to match the required format of the github-source connector ([here's an example](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-github/integration_tests/sample_config.json))

No I appreciate this is our API and I could do a conversion on the taskExecution side of things, but these routes are going to need to change anyway when we start doing UI-based configuration and such, so I want to keep the code simple and investment low

File: packages/hash/api/src/graphql/typeDefs/taskExecution.typedef.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/taskExecution.typedef.ts]: Have changed :) cheers for spotting

File: packages/hash/task-executor/src/airbyte/executor.ts
Comment (benwerner01) [inline packages/hash/task-executor/src/airbyte/executor.ts:165]: **Blocking:** this is never going to throw because an empty array `[]` is truthy. You probably want to do this instead:
```suggestion
    if (recordMessages.length > 0) {
      console.log(`There were ${recordMessages.length} records`);
      return recordMessages;
    } else {
      throw new Error("Message didn't contain any Airbyte Records");
    }
```

**Review by benwerner01 — CHANGES_REQUESTED** [review]
This is looking good 👍

I'm still trying to get this working locally, but in the meantime I've spotted a few minor blockers and some other optional suggestions/FYIs.

File: packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts]: Good point, Ben. I think the comment on the field refers to is not the field that is actually set though.
It says `alias of 'entityId'` but it is set as `id: this.entityVersionId,` in the model class.

File: packages/hash/task-executor/src/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/task-executor/src/index.ts:72]: `.json` I assume you mean :D

File: packages/hash/task-executor/src/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/task-executor/src/index.ts:8]: Adding a todo but just to put it out there again, keeping dev on the executor itself as minimal as possible. Going to swap to temporal or some other workflow executor soonish.

File: packages/hash/api/src/task-execution/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/task-execution/index.ts:63]: Regarding the Asana task, I did ask if we should block merging this PR to quickly extend the model class, but was informed it might be a little complicated. Our conversations about the graph right now and discovering types will address this, and will require us changing the model classes so I worry that us making a task to specifically support finding by name is wasted effort for now (as it will likely be made redundant by the URI work and structural queries).

File: packages/hash/api/src/task-execution/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/task-execution/index.ts:75]: This is also a leftover. I really can't remember why it was here but it was necessary before and wasn't properly inferred (it was `string | undefined | null`). Have removed cheers.

File: packages/hash/api/src/task-execution/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/task-execution/index.ts:97]: I think I'd previously had this code in a context with more than one step and used a transaction because of that. Have cleaned up

File: packages/hash/task-executor/nodemon.json
Comment (Alfred-Mountfield) [reply packages/hash/task-executor/nodemon.json:4]: Because I'm a dummy and I blindly copied the ts-node command from a different yarn script in the repo for consistency. 🙃

File: packages/hash/task-executor/README.md
Comment (Alfred-Mountfield) [reply packages/hash/task-executor/README.md:14]: Uh I don't think this should be blocking. I believe it's correct as is.

Python and pip are silly in that they don't let you keep track of top-level dependencies. So in npm land you put in your dependencies in your package.json and then your lock file is the fully qualified tree including all of the sub-dependencies.

If you run pip freeze it will output all installed packages, and you have no way of knowing which top-level ones there are.

Because of this, my Python setup is a little verbose but lets you keep track of your top-level dependencies in a requirements.txt file, **for updating** you then install those (which installs all of the sub-dependencies), and you freeze that into a requirements.lock file.
**For normal use**, you install from the lock file and have a consistent set of dependencies.

File: packages/hash/task-executor/src/airbyte/executor.ts
Comment (Alfred-Mountfield) [reply packages/hash/task-executor/src/airbyte/executor.ts:145]: These are the mounted volumes within the docker container. The `config.json` should be made in a temporary directory (see `writeToTempFile`) and shouldn't be located within the repo so adding it to `.gitignore` shouldn't do anything.

If that's not the case it's a bug or oversight.

File: packages/hash/task-executor/src/airbyte/executor.ts
Comment (Alfred-Mountfield) [reply packages/hash/task-executor/src/airbyte/executor.ts:165]: Good catch, I think my original error was in a try catch and wasn't about the length. (I was considering 0 records as being an acceptable state)

File: packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:197]: I'm a little worried about doing this. I'm not sure if this will overwhelm the DB right now (I would need to look up the implementation of promises and see if they're batched at all or if it ends up dependent on available threads and such but I suspect because it's IO that it will spam with possibly 1000s of requests in one go). Also I don't think the existingEntityChecker is set up for concurrency right now and without putting a little thought into the caching mechanism I'd not be comfortable adding this.

Don't get me wrong, the suggestion to parallelise is 100% valid, but I think we're solving it in the wrong place and way here. We need to think about the API better, especially around entityType discovery and mass data-insertion. We'll address that anyway soon so I don't think we need a specific asana task around it (and this code will be revisited and rewritten)

File: packages/hash/task-executor/src/airbyte/protocol.ts
Comment (Alfred-Mountfield) [reply packages/hash/task-executor/src/airbyte/protocol.ts:4]: Not quite, have updated. They decided to write their **JSON** schema in... YAML 😂

File: packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts:197]: Sounds good

File: packages/hash/task-executor/README.md
Comment (benwerner01) [reply packages/hash/task-executor/README.md:14]: Ah sorry I mistakenly thought you had renamed `requirements.txt` to `requirements.lock`, thanks for explaining why they both need to exist.

File: packages/hash/api/src/task-execution/index.ts
Comment (benwerner01) [reply packages/hash/api/src/task-execution/index.ts]: Makes sense, thanks for confirming

File: packages/hash/task-executor/src/airbyte/executor.ts
Comment (Alfred-Mountfield) [reply packages/hash/task-executor/src/airbyte/executor.ts:145]: Gotcha, apologies. I'd like to make a .gitignore in the `packages/hash/task-executor` folder because when this is deleted it'll be easy to make sure the ignore record is deleted too. 

I think we have some discussion about trying to keep things in a global .gitignore though (which would be a lot easier to miss) @kachkaev do you see problems with me making ``packages/hash/task-executor/.gitignore` for `catalog.json`?

Comment (benwerner01) [reply packages/hash/task-executor/src/airbyte/executor.ts:145]: I personally don't have an issue with having scoped `.gitignore` files for packages like this, but will defer to @kachkaev

File: packages/hash/task-executor/src/airbyte/protocol.ts
Comment (benwerner01) [reply packages/hash/task-executor/src/airbyte/protocol.ts:4]: gotcha

File: packages/hash/task-executor/README.md
Comment (Alfred-Mountfield) [reply packages/hash/task-executor/README.md:14]: It was more self-explanatory when there were dependencies. But we moved from Singer taps to AirByte connectors (I was doing dev on the same branch and rebased) and I'd already put in the code and explanation of how to use the Python stuff, I figured leave it there as it's useful, but now a little less clear just empty requirements files.

File: packages/hash/task-executor/src/airbyte/executor.ts
Comment (kachkaev) [reply packages/hash/task-executor/src/airbyte/executor.ts:145]: I don't mind having multiple `.gitignore` files, but from my experience this brings more complexities than benefits in a monorepo of our size. The cost of not removing some path from the root `.gitignore` is not that big compared to managing a fractal tree of configs. Deciding where to place a new file and what to put there is an extra human task that can be avoided. A single `.gitconfig` shows all the applied rules at a glance.

Up to you to decide. If we go for a new `.gitignore` file, let's chat about the rules of these files after this PR. We should have some sort of a system that helps us create the right number of `.gitignore` files with the right contents. Ideally, these rules should be simple and unambiguous. What I mean is that the number and the contents of `.gitignore` files should be a function of the today's state of the repo rather than its history.

Comment (Alfred-Mountfield) [reply packages/hash/task-executor/src/airbyte/executor.ts:145]: Makes sense, I thought there was a problem with the splitting the config due to some tooling not being able to resolve the child directories but maybe I'm misremembering and you were talking about something other than .gitignore.

I think for now I'll go ahead with adding one in this folder purely because we do plan on fully deleting the `task-executor` package and it's purposefully pretty self-contained.

Lets have a follow-up discussion as you say.

**Review by benwerner01 — APPROVED** [review]
Thanks Alfie 👌

**Pull request merged**
